00_Base/src/datasources/TokensDatasource.ts

```diff
@@ -0,0 +1,229 @@
+      this.logger.error('Error retrieving single token', error);
+      throw error;
+    }
+  }
+
+  async patchToken(partialToken: TokenDTO): Promise<TokenDTO> {
+    TokensValidators.validatePartialTokenForUniquenessRequiredFields(
+      partialToken,
+    );
+
+    const ocppAuths = await this.authorizationRepository.readAllByQuerystring({
+      idToken: partialToken.uid!,
+      type: OcpiTokensMapper.mapOcpiTokenTypeToOcppIdTokenType(
+        partialToken.type,
+      ),
+    });
+
+    if (ocppAuths.length === 0) {
+      throw new UnknownTokenException('Token not found in the database');
+    }
+
+    const existingOcpiToken = await this.tokensRepository.readOnlyOneByQuery({
+      where: {
+        authorization_id: {
+          [Op.in]: ocppAuths.map((ocppAuth) => ocppAuth.id),
+        },
+        country_code: partialToken.country_code,
+        party_id: partialToken.party_id,
+        type: partialToken.type,
+      },
+    });
+
+    if (!existingOcpiToken) {
+      throw new UnknownTokenException('Token not found in the database');
+    }
+
+    const updatedToken = await this.tokensRepository.updateByKey(
+      partialToken,
+      existingOcpiToken.id,
+    );
+
+    if (!updatedToken) {
+      throw new UnknownTokenException('Token not found in the database');
+    }
+    //
+    // const newOCPPAuth =
+    //   OCPITokensMapper.mapOcpiTokenToOcppAuthorization(updatedToken);
+    // newOCPPAuth.id = existingOCPIToken.id;
+    // const updatedOcppAuth =
+    //   await this.authorizationRepository.createOrUpdateByQuerystring(
+    //     newOCPPAuth,
+    //     {
+    //       idToken: partialToken.uid!,
+    //       type: OCPITokensMapper.mapOcpiTokenTypeToOcppIdTokenType(partialToken.type),
+    //     },
+    //   );
+    //
+    // if (!updatedOcppAuth) {
+    //   throw new Error('Could not save token');
+    // }
+
+    return new TokenDTO();
+  }
+
+  async updateToken(tokenDto: TokenDTO): Promise<TokenDTO> {
+    const mappedOcppAuth =
+      OcpiTokensMapper.mapOcpiTokenToOcppAuthorization(tokenDto);
+    const savedOcppAuth =
+      await this.authorizationRepository.createOrUpdateByQuerystring(
+        mappedOcppAuth,
+        {
+          idToken: tokenDto.uid,
+          type: OcpiTokensMapper.mapOcpiTokenTypeToOcppIdTokenType(
+            tokenDto.type,
+          ),
+        },
+      );
+
+    if (!savedOcppAuth) {
+      throw new UnknownTokenException(
+        'Authorization could not be created upon saving token',
+      );
+    }
+
+    try {
+      const ocpiToken = await this.tokensRepository.createOrUpdateOcpiToken(
+        savedOcppAuth?.id,
+        tokenDto,
+      );
+      return OcpiTokensMapper.toDto(savedOcppAuth!, ocpiToken);
+    } catch (error) {
+      this.logger.error('Error saving token', error);
+      throw error;
+    }
+  }
+
+  async updateBatchedTokens(tokens: TokenDTO[]): Promise<void> {
+    const batchFailedTokens: TokenDTO[] = [];
+
+    // TODO: do an actual bulk update
```

Comment:
1. I am not convinced this should be a bulk update. having it in a try catch loop seems more resilient to failures of individual bad tokens. This way we can investigate those cases and still have most updates go through.
2. Can we make it a Ticket instead of a code TODO if we decide it's worth it.

Reply:
Updated to not use batch